libraries/botbuilder-core/src/showTypingMiddleware.ts

```diff
@@ -20,104 +20,65 @@ import { TurnContext } from './turnContext';
          */
-    constructor(delay: number = 500, period: number = 2000) {
+    constructor(private readonly delay: number = 500, private readonly period: number = 2000) {
         if (delay < 0) {
             throw new Error('Delay must be greater than or equal to zero');
         }
 
         if (period <= 0) {
             throw new Error('Repeat period must be greater than zero');
         }
-
-        this.delay = delay;
-        this.period = period;
     }
 
     /** Implement middleware signature
          * @param context {TurnContext} An incoming TurnContext object.
          * @param next {function} The next delegate function.
          */
-    public async onTurn(context: TurnContext, next: () => Promise<void>): Promise<void> {
-
+    public async onTurn(context: TurnContext, next: () => Promise<void>) {
         let finished = false;
-        let hTimeout: any = undefined;
-
-        /**
-             * @param context TurnContext object representing incoming message.
-             * @param delay The initial delay before sending the first indicator.
-             * @param period How often to send the indicator after the first.
-             */
-        function startInterval(context: TurnContext, delay: number, period: number): void {
-            hTimeout = setTimeout(
-                async () => {
-                    if (!finished) {
-                        let typingActivity: Partial<Activity> = {
-                            type: ActivityTypes.Typing,
-                            relatesTo: context.activity.relatesTo
-                        };
-
-                        // Sending the Activity directly via the Adapter avoids other middleware and avoids setting the
-                        // responded flag. However this also requires that the conversation reference details are explicitly added.
-                        const conversationReference: Partial<ConversationReference> =
-                                TurnContext.getConversationReference(context.activity);
-                        typingActivity = TurnContext.applyConversationReference(typingActivity, conversationReference);
-
-                        await context.adapter.sendActivities(context, [typingActivity]);
-
-                        // Pass in period as the delay to repeat at an interval.
-                        startInterval(context, period, period);
-                    } else {
-                        // Do nothing! This turn is done and we don't want to continue sending typing indicators.
+        let timeout: ReturnType<typeof setTimeout>;
+
+        const scheduleIndicator = (delay = this.delay) => {
+            timeout = setTimeout(async () => {
+                if (!finished) {
+                    try {
+                        await this.sendTypingActivity(context);
+                    } catch (err) {
+                        await context.adapter.onTurnError(context, err);
```

Comment:
I'm not sure onTurnError is guaranteed to have a value?

Reply:
Good point, can we default it to `console.error(err)` or something of the sort? We can certainly only invoke it conditionally, I'm not sure how strict our parity requirement are around things like this.

Reply:
The latest [changes](https://github.com/microsoft/botbuilder-js/pull/2786/commits/bcfe26a4c7daa88bc3b16eb98b931cdef156933d#diff-5b9f058c68a2725e35cba9c53da4d2f4R52-R56) make it so that if the adapter doesn't have an onTurnError, it's going to result in an uncaught exception:

https://github.com/microsoft/botbuilder-js/blob/bcfe26a4c7daa88bc3b16eb98b931cdef156933d/libraries/botbuilder-core/src/showTypingMiddleware.ts#L51-L57

Theoretically speaking, those caring about uncaught exceptions will likely have an onTurnError, but others may still stray down the no `Adapter.onTurnError` path and end up with the same result described in issue #2527. 

___

The ShowTypingMiddleware constructors are the same between .NET, JS and Python; `constructor(delay: number, period: number)`. Ideally through either the constructor or public setters, we would have a settable `TelemetryClient` and/or an `ILogger` (.NET specific) that we could call when Adapter.OnTurnError is falsy.

Can you add a test for the missing `onTurnError` path?

Reply:
@stevengum this is intentional - (as far as I can tell) the current behavior for a Bot without an `onTurnError` handler is Promise rejection ([looking at this code](https://github.com/microsoft/botbuilder-js/blob/master/libraries/botbuilder-core/src/botAdapter.ts#L170)). Eventually (again, as far as I can tell), that Promise rejection is bubbled up as an UnhandledPromiseRejectionWarning. I'm not as well versed in the error handling across the board, so if it is indeed customary to mask/swallow errors then I'll relent, though I'll register some intuition that says masking errors is potentially problematic.

Reply:
Good point! This is almost certainly a non-scenario because we by default make all samples and generators/templates have an onTurnError. A developer would need to start from nothing to have non-negligible chances of not having an onTurnError.

The only reason we would potentially swallow the error is because it's (debatably) a non-critical operation from the bot. I'd prefer if we at least sent a trace to the App Insights indicating that an error occurred when trying to send a typing activity via a BotTelemetryClient, but that's out of scope and rethrowing the error is fine.